File: src/main/scala/org/xbery/artbeams/articles/domain/Article.scala
package org.xbery.artbeams.articles.domain

import java.util
import java.util.Date

import org.xbery.artbeams.common.assets.domain.{Asset, AssetAttributes, Validity, ValidityAsset}

/**
  * Article entity
  * @author Radek Beran
  */
@SerialVersionUID(-1L)
case class Article(
  override val common: AssetAttributes,
  override val validity: Validity,
  externalId: Option[String],
  slug: String,
  title: String,
  image: Option[String],
  perex: String,
  bodyMarkdown: String,
  body: String,
  keywords: String,
  showOnBlog: Boolean,
) extends Asset with ValidityAsset {

  def updatedWith(edited: EditedArticle, htmlBody: String, userId: String): Article = {
    this.copy(
      common = this.common.updatedWith(userId),
      validity = this.validity.updatedWith(edited),
      externalId = edited.externalId.flatMap(extId => if (extId.trim == "") None else Option(extId)),
      slug = edited.slug,
      title = edited.title,
      image = edited.image.flatMap(img => if (img.trim == "") None else Option(img)),
      perex = edited.perex,
      bodyMarkdown = edited.bodyMarkdown,
      body = htmlBody,
      keywords = edited.keywords,
      showOnBlog = edited.showOnBlog
    )
  }

  def toEdited(categories: util.List[String]): EditedArticle = {
    EditedArticle(
      this.id,
      this.externalId,
      this.slug,
      this.title,
      this.image,
      this.perex,
      this.bodyMarkdown,
      if (this.validFrom == null || this.validFrom == AssetAttributes.EmptyDate) { new Date() } else { new Date(this.validFrom.toEpochMilli()) },
      this.validTo.map(d => new Date(d.toEpochMilli())),
      this.keywords,
      this.showOnBlog,
      categories
    )
  }
}

object Article {
  final val CacheName = "articles"

  lazy val Empty = Article(
    AssetAttributes.Empty,
    Validity.Empty,
    None,
    "",
    "New article",
    None,
    "",
    "",
    "",
    "",
    true
  )

  lazy val EmptyEdited = Empty.toEdited(new util.ArrayList[String]())
}

File: src/main/kotlin/org/xbery/artbeams/articles/domain/Article.kt
package org.xbery.artbeams.articles.domain

import org.xbery.artbeams.common.assets.domain.Asset
import org.xbery.artbeams.common.assets.domain.AssetAttributes
import org.xbery.artbeams.common.assets.domain.Validity
import org.xbery.artbeams.common.assets.domain.ValidityAsset
import java.util.ArrayList
import java.util.Date

/**
 * Article entity
 *
 * @author Radek Beran
 */
data class Article(
    override val common: AssetAttributes,
    override val validity: Validity,
    val externalId: String?,
    val slug: String,
    val title: String,
    val image: String?,
    val perex: String,
    /** Body in editing format (e.g. in markdown). */
    val bodyEdited: String,
    val body: String,
    val keywords: String,
    val showOnBlog: Boolean,
    val editor: String,
) : Asset(),
    ValidityAsset {
    fun updatedWith(
        edited: EditedArticle,
        htmlBody: String,
        userId: String,
    ): Article =
        this.copy(
            common = this.common.updatedWith(userId),
            validity = this.validity.updatedWith(edited),
            externalId = (edited.externalId?.let { extId -> if (extId.trim() == "") null else extId }),
            slug = edited.slug,
            title = edited.title,
            image = (edited.image?.let { img -> if (img.trim() == "") null else img }),
            perex = edited.perex,
            bodyEdited = edited.bodyEdited,
            body = htmlBody,
            keywords = edited.keywords,
            showOnBlog = edited.showOnBlog,
            editor = edited.editor,
        )

    fun toEdited(categories: List<String>): EditedArticle {
        val validTo = this.validTo
        return EditedArticle(
            this.id,
            this.externalId,
            this.slug,
            this.title,
            this.image,
            null,
            this.perex,
            this.bodyEdited,
            this.editor,
            if (this.validFrom == AssetAttributes.EMPTY_DATE) {
                Date()
            } else {
                Date(this.validFrom.toEpochMilli())
            },
            if (validTo != null) {
                Date(validTo.toEpochMilli())
            } else {
                null
            },
            this.keywords,
            this.showOnBlog,
            categories,
        )
    }

    companion object {
        const val CacheName: String = "articles"
        val Empty: Article =
            Article(
                AssetAttributes.EMPTY,
                Validity.Empty,
                null,
                "",
                "New article",
                null,
                "",
                "",
                "",
                "",
                true,
                "markdown",
            )
        val EmptyEdited: EditedArticle = Empty.toEdited(ArrayList<String>())
    }
}
